File: docs/breez-sdk/snippets/rust/src/htlcs.rs
use anyhow::Result;
use breez_sdk_spark::*;
use log::info;

async fn send_htlc_payment(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: send-htlc-payment
    let payment_request = "<spark address>".to_string();
    // Set the amount you wish to pay the receiver
    let amount_sats = Some(50_000);
    let prepare_request = PrepareSendPaymentRequest {
        payment_request,
        amount: amount_sats,
        token_identifier: None,
        conversion_options: None,
        fee_policy: None,
    };
    let prepare_response = sdk.prepare_send_payment(prepare_request).await?;

    // If the fees are acceptable, continue to create the HTLC Payment
    if let SendPaymentMethod::SparkAddress { fee, .. } = prepare_response.payment_method {
        info!("Fees: {} sats", fee);
    }

    let preimage = "<32-byte unique preimage hex>";
    let preimage_bytes = hex::decode(preimage)?;
    let payment_hash_bytes = sha256::digest(preimage_bytes);
    let payment_hash = hex::encode(payment_hash_bytes);

    // Set the HTLC options
    let options = SendPaymentOptions::SparkAddress {
        htlc_options: Some(SparkHtlcOptions {
            payment_hash,
            expiry_duration_secs: 1000,
        }),
    };

    let request = SendPaymentRequest {
        prepare_response,
        options: Some(options),
        idempotency_key: None,
    };
    let send_response = sdk.send_payment(request).await?;
    let payment = send_response.payment;
    // ANCHOR_END: send-htlc-payment
    Ok(())
}

async fn receive_hodl_invoice_payment(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: receive-hodl-invoice-payment
    let preimage = "<32-byte unique preimage hex>";
    let preimage_bytes = hex::decode(preimage)?;
    let payment_hash_bytes = sha256::digest(preimage_bytes);
    let payment_hash = hex::encode(payment_hash_bytes);

    let response = sdk
        .receive_payment(ReceivePaymentRequest {
            payment_method: ReceivePaymentMethod::Bolt11Invoice {
                description: "HODL invoice".to_string(),
                amount_sats: Some(50_000),
                expiry_secs: None,
                payment_hash: Some(payment_hash),
            },
        })
        .await?;

    let invoice = response.payment_request;
    info!("HODL invoice: {invoice}");
    // ANCHOR_END: receive-hodl-invoice-payment
    Ok(())
}

async fn list_claimable_htlc_payments(sdk: &BreezSdk) -> Result<Vec<Payment>> {
    // ANCHOR: list-claimable-htlc-payments
    let request = ListPaymentsRequest {
        type_filter: Some(vec![PaymentType::Receive]),
        status_filter: Some(vec![PaymentStatus::Pending]),
        payment_details_filter: Some(vec![
            PaymentDetailsFilter::Spark {
                htlc_status: Some(vec![SparkHtlcStatus::WaitingForPreimage]),
                conversion_refund_needed: None,
            },
            PaymentDetailsFilter::Lightning {
                htlc_status: Some(vec![SparkHtlcStatus::WaitingForPreimage]),
            },
        ]),
        ..Default::default()
    };

    let response = sdk.list_payments(request).await?;
    let payments = response.payments;

    for payment in &payments {
        match &payment.details {
            Some(PaymentDetails::Spark {
                htlc_details: Some(htlc),
                ..
            }) => {
                info!("Spark HTLC expiry time: {}", htlc.expiry_time);
            }
            Some(PaymentDetails::Lightning {
                htlc_details: htlc,
                ..
            }) => {
                info!("Lightning HTLC expiry time: {}", htlc.expiry_time);
            }
            _ => {}
        }
    }
    // ANCHOR_END: list-claimable-htlc-payments
    Ok(payments)
}

async fn claim_htlc_payment(sdk: &BreezSdk) -> Result<Payment> {
    // ANCHOR: claim-htlc-payment
    let preimage = "<preimage hex>".to_string();
    let response = sdk
        .claim_htlc_payment(ClaimHtlcPaymentRequest { preimage })
        .await?;
    let payment = response.payment;
    // ANCHOR_END: claim-htlc-payment
    Ok(payment)
}

File: docs/breez-sdk/snippets/csharp/Htlcs.cs
using Breez.Sdk.Spark;

namespace BreezSdkSnippets
{
    class Htlcs
    {
        async Task SendHtlcPayment(BreezSdk sdk)
        {
            // ANCHOR: send-htlc-payment
            var paymentRequest = "<spark address>";
            // Set the amount you wish the pay the receiver
            ulong? amountSats = 50_000UL;
            var prepareRequest = new PrepareSendPaymentRequest(
                paymentRequest: paymentRequest,
                amount: amountSats,
                tokenIdentifier: null,
                conversionOptions: null,
                feePolicy: null
            );
            var prepareResponse = await sdk.PrepareSendPayment(request: prepareRequest);

            // If the fees are acceptable, continue to create the HTLC Payment
            if (prepareResponse.paymentMethod is SendPaymentMethod.SparkAddress sparkMethod)
            {
                var fee = sparkMethod.fee;
                Console.WriteLine($"Fees: {fee} sats");
            }

            var preimage = "<32-byte unique preimage hex>";
            var preimageBytes = Convert.FromHexString(preimage);
            var paymentHashBytes = System.Security.Cryptography.SHA256.HashData(preimageBytes);
            var paymentHash = Convert.ToHexString(paymentHashBytes).ToLower();

            // Set the HTLC options
            var options = new SendPaymentOptions.SparkAddress(
                htlcOptions: new SparkHtlcOptions(
                    paymentHash: paymentHash,
                    expiryDurationSecs: 1000
                )
            );

            var request = new SendPaymentRequest(
                prepareResponse: prepareResponse,
                options: options
            );
            var sendResponse = await sdk.SendPayment(request: request);
            var payment = sendResponse.payment;
            // ANCHOR_END: send-htlc-payment
        }

        async Task ReceiveHodlInvoicePayment(BreezSdk sdk)
        {
            // ANCHOR: receive-hodl-invoice-payment
            var preimage = "<32-byte unique preimage hex>";
            var preimageBytes = Convert.FromHexString(preimage);
            var paymentHashBytes = System.Security.Cryptography.SHA256.HashData(preimageBytes);
            var paymentHash = Convert.ToHexString(paymentHashBytes).ToLower();

            var response = await sdk.ReceivePayment(
                request: new ReceivePaymentRequest(
                    paymentMethod: new ReceivePaymentMethod.Bolt11Invoice(
                        description: "HODL invoice",
                        amountSats: 50_000UL,
                        expirySecs: null,
                        paymentHash: paymentHash
                    )
                )
            );

            var invoice = response.paymentRequest;
            Console.WriteLine($"HODL invoice: {invoice}");
            // ANCHOR_END: receive-hodl-invoice-payment
        }

        async Task ListClaimableHtlcPayments(BreezSdk sdk)
        {
            // ANCHOR: list-claimable-htlc-payments
            var request = new ListPaymentsRequest(
                typeFilter: new List<PaymentType> { PaymentType.Receive },
                statusFilter: new List<PaymentStatus> { PaymentStatus.Pending },
                paymentDetailsFilter: new List<PaymentDetailsFilter> {
                    new PaymentDetailsFilter.Spark(
                        htlcStatus: new List<SparkHtlcStatus> {
                            SparkHtlcStatus.WaitingForPreimage
                        },
                        conversionRefundNeeded: null
                    ),
                    new PaymentDetailsFilter.Lightning(
                        htlcStatus: new List<SparkHtlcStatus> {
                            SparkHtlcStatus.WaitingForPreimage
                        }
                    )
                }
            );

            var response = await sdk.ListPayments(request: request);
            var payments = response.payments;

            foreach (var payment in payments)
            {
                if (payment.details is PaymentDetails.Spark sparkDetails && sparkDetails.htlcDetails != null)
                {
                    Console.WriteLine($"Spark HTLC expiry time: {sparkDetails.htlcDetails.expiryTime}");
                }
                else if (payment.details is PaymentDetails.Lightning lightningDetails)
                {
                    Console.WriteLine($"Lightning HTLC expiry time: {lightningDetails.htlcDetails.expiryTime}");
                }
            }
            // ANCHOR_END: list-claimable-htlc-payments
        }

        async Task ClaimHtlcPayment(BreezSdk sdk)
        {
            // ANCHOR: claim-htlc-payment
            var preimage = "<preimage hex>";
            var response = await sdk.ClaimHtlcPayment(
                request: new ClaimHtlcPaymentRequest(preimage: preimage)
            );
            var payment = response.payment;
            // ANCHOR_END: claim-htlc-payment
        }
    }
}
